### ISSUE TITLE
Exception + VM hangs when shutting down via "reboot" API if network connections exist or have existed

### ISSUE THREAD

**Issue by lgoldstein** [cloudius-systems/osv#495]
I am using the _reboot_ command (same one that is the subject of issue #484) in order to shut down the VM. If I do it while there are no other active open incoming/outgoing network connections then everything is fine. However, if I do it during an active network connection (SSH in this case), the exception below occurs and the VM "hangs" - i.e., it does not complete the power down cycle.

![capture2](https://cloud.githubusercontent.com/assets/1436205/4231699/73fa46b0-3994-11e4-97af-f4a124e50fd0.PNG)
- This is on _Vmware Workstation_

**Comment by nyh** [comment]
Hi Lyor, could you provide me a few more hints on how I can reproduce this crash? How do you build an image which contains SSH (sorry, I've lost track of the various variants of OSv's shells....)? How do you cause it to reboot (you type a command, or what?)? Does this bug only happen on VMware or also on KVM?

One thing surprising me about this crash, if it is indeed consistent, is how come you always see this close() call in Java code during the reboot. A reboot doesn't normally close anything (it's a single "outb" instruction  which should cause the hypervisor to stop this virtual core and (eventually) the other virtual cores too).

**Comment by lgoldstein** [comment]
I am running Apache SSHD (Java) and create a connection to it. Then I disconnect and try to shutdown the system - which is when the exception occurs. I believe it has something to do with the *epoll...* which is used by the Apache SSHD to do asynchronous socket I/O

**Comment by nyh** [comment]
On Mon, Feb 2, 2015 at 3:31 PM, Lyor Goldstein <notifications@github.com>
wrote:

> I am running Apache SSHD (Java) and create a connection to it. Then I
> disconnect and try to shutdown the system
>

How do you "try to shutdown the system" - is that a command on the console?
A network connection? Do you try to "shutdown" the system, or "reboot" it
(as you said earlier), or what exactly?

Sorry for all the nagging :-)


> - which is when the exception occurs. I believe it has something to do
> with the *epoll...* which is used by the Apache SSHD to do asynchronous
> socket I/O
>

It definitely looks this way, but I'm trying to figure out why the reboot
triggers this problem.


-- 
Nadav Har'El
nyh@cloudius-systems.com

**Comment by lgoldstein** [comment]
I am calling the [reboot](http://www.unix.com/man-page/freebsd/2/reboot/) syscall using either *RB_AUTOBOOT* or *RB_POWER_OFF* - according to what I want to do (shutdown or reboot)

**Comment by lgoldstein** [comment]
Here are updated screen captures from the latest failure:
![capture-1](https://cloud.githubusercontent.com/assets/1436205/6248377/a43357b0-b787-11e4-9616-0d3ef1c006e6.JPG)
![capture-2](https://cloud.githubusercontent.com/assets/1436205/6248380/b59fcfd8-b787-11e4-8974-9115edf7c345.JPG)

**Comment by lgoldstein** [comment]
Hi guys, any progress on this issue - it still happens pretty much the same way described in this issue...

**Comment by nyh** [comment]
Hi Lyor, I don't think there was any progress on this. I never tried to reproduce this issue myself.

I'm not surprised there is a bug in the epoll code - it is fairly complex code with a lot of subtle corner cases. However, what really puzzles me, is how osv::poweroff() or osv::reboot() can possibly disturb it. These commands do not attempt any cleanups of any kind, so how can they bother any epoll code running in parallel?

**Comment by lgoldstein** [comment]
@nyh I wish I knew - if as you say _These commands do not attempt any cleanups of any kind_, then how do you explain it ? As far as reproducing this issue, it occurs fairly regularly on my image - I can give you an OVA or VMDK + VMX file of my machine with simple instructions how to reproduce it. It is not a high priority issue (yet), but would be nice to have it fixed...
